dotCMS/src/main/java/com/dotcms/cache/VanityUrlCacheImpl.java

```diff
@@ -67,20 +74,74 @@ public void remove(final Contentlet vanityURL) {
 
         try {
 
+            String siteId = vanityURL.getStringProperty(VanityUrlContentType.SITE_FIELD_VAR);
+
             /*
-            First get the records we want to remove from secondary cache group and
-            delete it from the primary cache group, we do this in order to avoid to
-            flush the primary group.
+            First get the records we want to remove from the secondary cache group and
+            remove each of those records from the primary cache group, we do this in order to avoid to
+            flush completely the primary group.
              */
-            Set<CachedVanityUrl> cachedVanityUrls = this
-                    .getCachedVanityUrls(VanityUrlUtil.sanitizeSecondCachedKey(vanityURL));
-            for (CachedVanityUrl toRemove : cachedVanityUrls) {
-                this.remove(VanityUrlUtil.sanitizeKey(toRemove.getSiteId(), toRemove.getUrl(),
-                        toRemove.getLanguageId()));
+            Set<CachedVanityUrl> secondaryCachedVanities = this
+                    .getCachedVanityUrls(VanityUrlUtil.sanitizeSecondCacheKey(vanityURL));
+            if (null != secondaryCachedVanities) {
+                for (CachedVanityUrl toRemove : secondaryCachedVanities) {
+                    //Now remove from the primary group
+                    this.remove(VanityUrlUtil.sanitizeKey(toRemove.getSiteId(), toRemove.getUrl(),
+                            toRemove.getLanguageId()));
+                }
+            }
+
+            /*
+            Now we want to clean up the 404 cache records and only touch other sites if we
+            are trying to invalidate a SYSTEM_HOST VanityURL, those vanities affects all the sites
+             */
+            if (Host.SYSTEM_HOST.equals(siteId)) {
+
+                //Get all the sites related to these cached VanityURLs
+                Set<String> sites = getSiteIds();
+                if (null != sites) {
+                    for (String site : sites) {
+
+                        if (!Host.SYSTEM_HOST.equals(site)) {
```

Comment:
I would definitely pay attention to warning-errors reported by jenkins-dotcms like these. Avoiding this levels of nested "if" blocks helps tremendously in code reading/maintainability